File: java/BinaryTreeInorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode node = root;
        List<Integer> result = new ArrayList<Integer>();
        
        while (stack.isEmpty() == false || node != null) {
            if (node != null) {
                stack.push(node);
                node = node.left;
            } else {
                TreeNode peak = stack.pop();
                result.add(peak.val);
                node = peak.right;
            }
        }
        
        return result;
    }
}

File: archive/python/BinaryTreeInorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def inorderTraversal(self, root):
        stack = []
        node = root
        result = []
        while len(stack) > 0 or node:
            if node:
                stack.append(node)
                node = node.left
            else:
                peak = stack.pop()
                result.append(peak.val)
                node = peak.right
        return result
